flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/FlinkKubernetesCustomCli.java

```diff
@@ -0,0 +1,377 @@
+				commandLine.getOptionValue(imageOption.getOpt()));
+		}
+
+		if (commandLine.hasOption(jobManagerMemeoryOption.getOpt())) {
+			String jmMemoryVal = commandLine.getOptionValue(jobManagerMemeoryOption.getOpt());
+			if (!MemorySize.MemoryUnit.hasUnit(jmMemoryVal)) {
+				jmMemoryVal += "m";
+			}
+			effectiveConfiguration.setString(JobManagerOptions.JOB_MANAGER_HEAP_MEMORY, jmMemoryVal);
+		}
+
+		if (commandLine.hasOption(taskManagerMemeoryOption.getOpt())) {
+			String tmMemoryVal = commandLine.getOptionValue(taskManagerMemeoryOption.getOpt());
+			if (!MemorySize.MemoryUnit.hasUnit(tmMemoryVal)) {
+				tmMemoryVal += "m";
+			}
+			effectiveConfiguration.setString(TaskManagerOptions.TASK_MANAGER_HEAP_MEMORY, tmMemoryVal);
+		}
+
+		if (commandLine.hasOption(taskManagerSlotsOption.getOpt())) {
+			effectiveConfiguration.setInteger(TaskManagerOptions.NUM_TASK_SLOTS,
+				Integer.parseInt(commandLine.getOptionValue(taskManagerSlotsOption.getOpt())));
+		}
+
+		Configuration dynamicConf = ConfigurationUtils.createConfiguration(
+			commandLine.getOptionProperties(dynamicPropertiesOption.getOpt()));
+		for (String key : dynamicConf.keySet()) {
+			effectiveConfiguration.setString(key, dynamicConf.getString(key, ""));
+		}
+
+		StringBuilder entryPointClassArgs = new StringBuilder();
+		if (commandLine.hasOption(jobClassOption.getOpt())) {
+			entryPointClassArgs.append(" --")
+				.append(jobClassOption.getLongOpt())
+				.append(" ")
+				.append(commandLine.getOptionValue(jobClassOption.getOpt()));
+		}
+
+		if (commandLine.hasOption(jobIdOption.getOpt())) {
+			entryPointClassArgs.append(" --")
+				.append(jobIdOption.getLongOpt())
+				.append(" ")
+				.append(commandLine.getOptionValue(jobIdOption.getOpt()));
+		}
+		if (!entryPointClassArgs.toString().isEmpty()) {
+			effectiveConfiguration.setString(FlinkMasterDeploymentDecorator.ENTRY_POINT_CLASS_ARGS,
+				entryPointClassArgs.toString());
+		}
+
+		return effectiveConfiguration;
+	}
+
+	private int run(String[] args) throws CliArgsException, FlinkException {
+		final CommandLine cmd = parseCommandLineOptions(args, true);
+
+		if (cmd.hasOption(HELP_OPTION.getOpt())) {
+			printUsage();
+			return 0;
+		}
+
+		final Configuration configuration = applyCommandLineOptionsToConfiguration(cmd);
+		final ClusterClientFactory<String> kubernetesClusterClientFactory =
+			clusterClientServiceLoader.getClusterClientFactory(configuration);
+
+		final ClusterDescriptor<String> kubernetesClusterDescriptor =
+			kubernetesClusterClientFactory.createClusterDescriptor(configuration);
+
+		try {
+			final ClusterClient<String> clusterClient;
+			String clusterId = kubernetesClusterClientFactory.getClusterId(configuration);
+			boolean detached = cmd.hasOption(DETACHED_OPTION.getOpt());
+			FlinkKubeClient kubeClient = KubeClientFactory.fromConfiguration(configuration);
+
+			// Retrieve or create a session cluster.
+			if (clusterId != null && kubeClient.getInternalService(clusterId) != null) {
```

Comment:
Should user be notified if their specified cluster does not exist?

Reply:
The code path is for `kubernetes-session.sh`. If the cluster exists, we will retrieve and attach to it. Otherwise, we create a new one. It just has the same behavior with `yarn-session.sh`.

`./bin/kubernetes-session.sh -d -id flink-native-k8s-session-1`